src/core_write.cpp

```diff
@@ -169,17 +172,22 @@ void ScriptPubKeyToUniv(const CScript& scriptPubKey,
         return;
     }
 
-    out.pushKV("reqSigs", nRequired);
+    if (ExtractDestination(scriptPubKey, address)) {
```

Comment:
See my earlier edit here: https://github.com/bitcoin/bitcoin/pull/20286#discussion_r549429414

For "address" we shouldn't report anything for P2PK outputs either, as the address that's reported currently is really the P2PKH one for the same pubkey (and sending to that P2PKH address would have a distinct scriptPubKey). This is due to historical reasons where addresses used to also be "pubkey id" (that's the same reason as why "addresses" were reported for bare multisig; those addresses in modern meaning have nothing to do with that multisig address).

Can you add a `&& GetScriptForDestination(address) == scriptPubKey`? That would guarantee that whatever address is extracted exactly matches the output.

Reply:
I pushed a commit as you suggest: https://github.com/bitcoin/bitcoin/pull/20286/commits/97fbaffc4aea265d638895531dc3b3c84669c561

I'm a little confused though. Aren't we already not reporting anything for P2PK bc we return here https://github.com/bitcoin/bitcoin/blob/master/src/core_write.cpp#L167, if `type == TxoutType::PUBKEY`?

Also, I want to know if this problem could affect us anywhere else. It seems `ExtractDestination` already handles this https://github.com/bitcoin/bitcoin/blob/master/src/core_write.cpp#L151 (only pushes an `address` if `type != TxoutType::PUBKEY`) -- but if not, `rawtransaction.cpp` invokes `ScriptToUniv` which may need this check as well

Last, I'm trying to figure out if anything needs to be done in `rpcwallet`, here: https://github.com/bitcoin/bitcoin/blob/master/src/wallet/rpcwallet.cpp#L3669

Reply:
> I'm a little confused though. Aren't we already not reporting anything for P2PK bc we return here https://github.com/bitcoin/bitcoin/blob/master/src/core_write.cpp#L167, if `type == TxoutType::PUBKEY`?

You're right, this change isn't needed. It seems the specific P2PK case was already addressed earlier.

> Also, I want to know if this problem could affect us anywhere else. It seems `ExtractDestination` already handles this https://github.com/bitcoin/bitcoin/blob/master/src/core_write.cpp#L151 (only pushes an `address` if `type != TxoutType::PUBKEY`) -- but if not, `rawtransaction.cpp` invokes `ScriptToUniv` which may need this check as well

The code you're linking to is `ScriptToUniv`, not `ExtractDestination`. `ExtractDestination` does produce output for P2PK outputs. It makes sense to see what other call sites may be relying on this P2PK behavior though; if there aren't any, or they are all bogus, it may be easier to just drop support for P2PK in `ExtractDestinations`.

> Last, I'm trying to figure out if anything needs to be done in `rpcwallet`, here: https://github.com/bitcoin/bitcoin/blob/master/src/wallet/rpcwallet.cpp#L3669

That looks like a bug, but a hard to trigger one. You'd need a P2PK script inside a P2SH or P2WSH script in a wallet for that.

Reply:
OK, I'll go ahead and revert the commit since it isn't needed.

> The code you're linking to is `ScriptToUniv`, not `ExtractDestination`. `ExtractDestination` does produce output for P2PK outputs. It makes sense to see what other call sites may be relying on this P2PK behavior though; if there aren't any, or they are all bogus, it may be easier to just drop support for P2PK in `ExtractDestinations`.

Ah yeah good point. Well I'll read through the code and try to figure out if an issue needs to be created. I guess if `ExtractDestination(s)` dropped support for `P2PK` it may also fix`rpcwallet` (if it does turn out to be a bug) https://github.com/bitcoin/bitcoin/blob/master/src/wallet/rpcwallet.cpp#L3669. It'll take me some time to fully understand all this and test some things out